usr/lib/tpm_stdll/tpm_openssl.c

```diff
@@ -73,27 +75,40 @@ RSA *openssl_gen_key()
 #ifdef OLDER_OPENSSL
     rsa = RSA_generate_key(2048, 65537, NULL, NULL);
     if (rsa == NULL) {
+        fprintf(stderr, "Error generating user's RSA key\n");
+        ERR_load_crypto_strings();
+        ERR_print_errors_fp(stderr);
+        goto err;
+    }
+
+    rc = RSA_check_key(rsa);
 #else
     bne = BN_new();
     rc = BN_set_word(bne, 65537);
     if (!rc) {
         fprintf(stderr, "Error generating bne\n");
         ERR_load_crypto_strings();
         ERR_print_errors_fp(stderr);
-        return NULL;
+        goto err;
     }
 
-    rsa = RSA_new();
-    rc = RSA_generate_key_ex(rsa, 2048, bne, NULL);
-    if (!rc) {
-#endif
+    ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+    if (ctx == NULL)
+        goto err;
+
+    if (EVP_PKEY_keygen_init(ctx) <= 0
+        || EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, 2048) <= 0
+        || EVP_PKEY_CTX_set_rsa_keygen_pubexp(ctx, bne) <= 0
+        || EVP_PKEY_keygen(ctx, &pkey) <= 0
+        || (rsa = EVP_PKEY_get1_RSA(pkey)) == NULL) {
         fprintf(stderr, "Error generating user's RSA key\n");
         ERR_load_crypto_strings();
         ERR_print_errors_fp(stderr);
-        return NULL;
+        goto err;
     }
 
-    rc = RSA_check_key(rsa);
+    rc = EVP_PKEY_check(ctx) == 1 ? 1 : 0;
```

Comment:
EVP_PKEY_check is also only available since OpenSSL 1.1.1.

The #ifdef OLDER_OPENSSL checks for the wrong OpenSSL version for this specific function:

```
#if OPENSSL_VERSION_NUMBER < 0x10100000L
#define OLDER_OPENSSL
#endif
```

Reply:
yes. i should not use the already existing OLDER_OPENSSL macro since it checks for 1.1.0 vs 1.0.2. I should use a new 1.1.1 macro for all of my changes since the apis are not deprectaed in 1.1.0 anyway so theres no need to change that.

Reply:
done